### PR TITLE
Fix: Automatically propagate user permission updates

### LINKED ISSUE #3635: [Permissions - QA] Deleted Permissions not reflected in UI until hard refresh
## Steps to reproduce

1. User 1 deletes the Owner permission of User 2.
2. User 2 attempts to make a simple payment using a decision method.
3. Perform a hard refresh on the page.

## Expected behaviour

* Deleted permissions should disappear from the UI immediately after the deletion action without needing a hard refresh.

## Actual behaviour

* Deleted permissions remain visible in the UI until a hard refresh is performed.

image

### PR DESCRIPTION
**Pull request by rumzledz**
## Description

> [!NOTE]
> Please note that there is an existing issue where the Permissions table on the Completed Action form is not correctly showing the permissions that were removed from the user after a Reputation or Multi-Sig motion has successfully been supported/approved and finalised. This is kind of a complex issue and **it will not be addressed on this PR**. 
>
> There is also an existing issue with reputation motions where the Finalise button has the possibility of getting stuck in a Pending state: https://github.com/JoinColony/colonyCDapp/issues/3859. You may or may not come across it but just know that this PR has nothing to do with it.

This PR makes use of subscriptions to automatically refresh member permissions information whenever it's updated.

**Out of scope yet relevant and manageable fixes**: You should also now be able to see the permissions that are marked for removal on the Completed Action component. BUT please note that this fix is only **limited to actions done via Permissions** as well as with motions that **have not been successfully supported/approved and finalised.** Please bear this in mind during testing 🙏 

Lastly, I did a wee refactor for setting the user data in our `AppContextProvider`.

add-remove-permissions

## Testing

> [!NOTE]
> Make sure that Fry doesn't not have any permissions to begin with

> [!IMPORTANT]
> - Please do not reload the page in between steps
> - Uninstall the Reputation and Multi-Sig extensions

1. Open an incognito window
2. Connect Fry's wallet
3. Bring up the Simple Payment form
4. Verify that you see this error banner:

image

5. In the non-incognito window, connect Leela's wallet
6. Bring up the Manage Permissions form
7. Give Fry Payer permissions in General
8. Click the incognito window and verify that the error banner has disappeared
9. Click the Decision Method field and verify that you see Permissions in the options list
10. Click the non-incognito window
11. Remove Fry's permissions from General
12. Click the incognito window
13. Verify that you see the error banner again:

image

**New stuff** ✨

* Added a new util `removeObjectFields` which allows you to remove a set of fields from a given object. It's strictly typed so you should be able to see auto-completed fields as you're selecting them:

image

Resolves #3635

### CODE REVIEW

**Review by rdig — DISMISSED** [review]
Error banner correctly changes state based on the updated permissions.

Note that I've not refreshed either page in between any of the actions

Screenshot from 2024-12-11 15-57-47
Screenshot from 2024-12-11 15-58-20
Screenshot from 2024-12-11 15-58-26
Screenshot from 2024-12-11 15-59-37
Screenshot from 2024-12-11 15-59-40

File: src/graphql/generated.ts
Comment (mmioana) [inline src/graphql/generated.ts:1]: @rumzledz please try running codegen again, as we shouldn't be removing the `expenditureFunding` and `expenditureId` for the `ColonyMultiSig` model

**Review by iamsamgibbs — APPROVED** [review]
Nice job, this works very nicely!

Before:

Screenshot 2024-12-12 at 10 08 22


https://github.com/user-attachments/assets/a459ed0d-3de6-476d-8a09-c4919fc3f599

File: src/graphql/generated.ts
Comment (rumzledz) [reply src/graphql/generated.ts:1]: Apologies for this @mmioana , I'm not sure what happened. I just ran it and it generated these for me 😓 But I shall generate it again 👌

Comment (rumzledz) [reply src/graphql/generated.ts:1]: All right pushed! 🚀 @mmioana

**Review by iamsamgibbs — APPROVED** [review]

**Review by rdig — COMMENTED** [review]
Re-approving. Good to go still

**Review by rdig — APPROVED** [review]

**Review by mmioana — APPROVED** [review]
Still working as expected @rumzledz ✨ 

Screenshot 2024-12-13 at 10 16 00
Screenshot 2024-12-13 at 10 17 06
Screenshot 2024-12-13 at 10 17 16
Screenshot 2024-12-13 at 10 17 26
Screenshot 2024-12-13 at 10 17 54
Screenshot 2024-12-13 at 10 18 05

Good to go! 🚢

**Pull request merged**